format.js

```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: 2022 Andre 'Staltz' Medeiros
+//
+// SPDX-License-Identifier: LGPL-3.0-only
+
+const BFE = require('ssb-bfe')
+const Ref = require('ssb-ref')
+const Uri = require('ssb-uri2')
+const { box, unboxKey, unboxBody } = require('envelope-js')
+const { SecretKey } = require('ssb-private-group-keys')
+const makeKeysManager = require('./keys')
+
+const name = 'box2'
+
+let _keysManager = null
+let _selfId = null
+
+function setup(config, cb) {
+  _selfId = config.keys.id
+  _keysManager = makeKeysManager(config)
+  // FIXME: load ssb-keyring here
+  cb()
+}
+
+function _isGroup(recp) {
+  return _keysManager.groupKey(recp) !== undefined
+}
+
+function _isFeed(recp) {
+  return (
+    Ref.isFeed(recp) ||
+    Uri.isFeedSSBURI(recp) ||
+    Uri.isBendyButtV1FeedSSBURI(recp) ||
+    Uri.isButtwooV1FeedSSBURI(recp)
+  )
+}
+
+function addOwnDMKey(key) {
+  _keysManager.addOwnDMKey(key)
+}
+
+function addGroupKey(id, key) {
+  _keysManager.addGroupKey(id, key)
+}
+
+function encrypt(plaintextBuf, opts) {
+  const recps = opts.recps
+  const selfId = opts.keys.id
+
+  const validRecps = recps
+    .filter((recp) => typeof recp === 'string')
+    .filter((recp) => recp === selfId || _isGroup(recp) || _isFeed(recp))
+
+  if (validRecps.length === 0) {
+    // prettier-ignore
+    throw new Error(`no box2 keys found for recipients: ${recps}`)
+  }
+  if (validRecps.length > 16) {
+    // prettier-ignore
+    throw new Error(`private-group spec allows maximum 16 slots, but you've tried to use ${validRecps.length}`)
+  }
+  // FIXME: move these validations to ssb-groups
+  // if (validRecps.filter(isGroup).length === 0) {
+  //   // prettier-ignore
+  //   throw new Error(`no group keys found among recipients: ${recps}`)
+  // }
+  // if (!isGroup(validRecps[0])) {
+  //   // prettier-ignore
+  //   throw new Error(`first recipient must be a group, but you've tried to use ${validRecps[0]}`)
+  // }
+  const groupRecpsCount = validRecps.filter(_isGroup).length
+  if (groupRecpsCount > 1) {
+    // prettier-ignore
+    throw new Error(`private-group spec only supports one group recipient, but you've tried to use ${groupRecpsCount}`)
+  }
+
+  const encryptionKeys = validRecps.reduce((acc, recp) => {
+    if (recp === selfId) return [...acc, ..._keysManager.ownDMKeys()]
+    else if (_isGroup(recp)) return [...acc, _keysManager.groupKey(recp)]
+    else if (_isFeed(recp)) return [...acc, _keysManager.sharedDMKey(recp)]
```

Comment:
this sharedDMKey is gonna depend on who the "author feed" is.

Reply:
selfId...

This `encrypt` is gonna be per-feed I guess... so is this are you registering this format per-feed?

Reply:
Need to get my head around how this fits into the publish pipeline

Reply:
Assume it will be something like

```js
ssb.db.createAs(myChessKeys, { content, encryptionFormat: 'box2' }, cb)
```

Reply:
Yes, but not `createAs`:

```js
ssb.db.create({ keys: myChessKeys, content, encryptionFormat: 'box2' }, cb)
```

Reply:
The "author feed" is given by `opts.keys.id` and this refers to the keys that are signing/creating the message.